### PR TITLE
build dir as order only prerequisite, docker rmi deduplicated, docker…

### PR DESCRIPTION
**Pull request by michalpiasecki1**
docker build and docker tag deduplicated
build dir as order only prerequisite, 
docker rmi deduplicated

Resolves: https://github.com/golemfactory/concent-deployment/issues/194, https://github.com/golemfactory/concent-deployment/issues/193

### CODE REVIEW

**Comment by kbeker** [comment]
Please if you adding PR write which issue you resolve.
e.g. 
```
Resolves #issue_number
```
Also please keep up with ours branch naming style :) : https://github.com/Code-Poets/coding-style/issues/13

**Comment by cameel** [comment]
> Resolves: #194, #193

This does not work like this. Please see [Closing issues using keywords](https://help.github.com/articles/closing-issues-using-keywords).

> To close multiple issues, preface each issue reference with one of the above keywords. You must use the keyword before each issue you reference for the keyword to work.
>
> For example, `This closes #34, closes #23, and closes example_user/example_repo#42` would close issues #34 and #23 in the same repository, and issue #42 in the "example_user/example_repo" repository.

File: containers/Makefile [hunk 1]
```diff
@@ -59,42 +64,30 @@ $(info IMAGE_PREFIX=$(IMAGE_PREFIX))
...
+	$(DOCKER_TAG)
 
-concent-api-clean:
-	docker rmi $(IMAGE_PREFIX)concent-api:$(IMAGE_VERSION) --force
+docker-clean: nginx-proxy-rmi nginx-storage-rmi postgresql-rmi concent-api-rmi  verifier-rmi
 
-verifier-clean:
-	docker rmi $(IMAGE_PREFIX)verifier:$(IMAGE_VERSION) --force
+%-rmi:
+	docker rmi $(IMAGE_PREFIX)$*:$(IMAGE_VERSION) --force
```
Comment (cameel) [inline containers/Makefile]: Any target that does not really create a file with the corresponding name must be a [Phony target](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html).

File: containers/Makefile [hunk 2]
```diff
@@ -59,42 +64,30 @@ $(info IMAGE_PREFIX=$(IMAGE_PREFIX))
...
 
-postgresql-clean:
-	docker rmi $(IMAGE_PREFIX)postgresql:$(IMAGE_VERSION) --force
+verifier: $(VERIFIER_SOURCE) build/repositories/golem/ build/verifier/golem/scripts/ build/verifier/golem/imgverifier-requirements.txt | build/
+	$(DOCKER_BUILD)
+	$(DOCKER_TAG)
 
-concent-api-clean:
-	docker rmi $(IMAGE_PREFIX)concent-api:$(IMAGE_VERSION) --force
+docker-clean: nginx-proxy-rmi nginx-storage-rmi postgresql-rmi concent-api-rmi  verifier-rmi
```
Comment (cameel) [inline containers/Makefile]: Please keep whitespace consistent.

File: containers/Makefile (hunk 1 above)
Comment (cameel) [inline containers/Makefile]: The change to the `clean` targets should be in a separate commit from the change to the image targets. They're independent of each other (one does not really require the other to work) so they can be separated easily. And keeping them in the same commit makes review harder - just look how changes from one are mixed with changes from the other in the diff.

File: containers/Makefile [hunk 3]
```diff
@@ -59,42 +64,30 @@ $(info IMAGE_PREFIX=$(IMAGE_PREFIX))
...
-	docker build --tag $(IMAGE_PREFIX)verifier:$(IMAGE_VERSION) build/verifier/
-	docker tag $(IMAGE_PREFIX)verifier:$(IMAGE_VERSION) $(IMAGE_PREFIX)verifier:latest
+nginx-proxy: $(NGINX_PROXY_SOURCE) build/nginx-proxy/concent-api-assets.tar build/nginx-proxy/concent-api-docs.tar | build/
+	$(DOCKER_BUILD)
+	$(DOCKER_TAG)
 
-docker-clean: nginx-proxy-clean concent-api-clean postgresql-clean verifier-clean
+nginx-storage: $(NGINX_STORAGE_SOURCE) | build/
+	$(DOCKER_BUILD)
+	$(DOCKER_TAG)
```
Comment (cameel) [inline containers/Makefile]: This does reduce duplication a bit but has two downsides:
- You still have to repeat those three lines for every image.
- The variables used like this look quite strange and opaque. This works but putting whole bash lines into variables feels kinda like you're trying really hard to create a function. I'd rather create a helper bash script and call it than do it this way.

I was actually thinking about doing it like we did in one of our earlier projects. First define a list of target names and then automatically declare the targets with a [Static Pattern Rule](https://www.gnu.org/software/make/manual/html_node/Static-Pattern.html#Static-Pattern). Prerequisites can be added using a special intermediate target.

Here's an incomplete example:

``` make
IMAGE_TARGETS := \
    nginx-proxy  \
    concent-api  \
    postgresql

nginx-proxy!dependencies:                    \
    $(NGINX_PROXY_SOURCE)                    \
    build/nginx-proxy/concent-api-assets.tar \
    build/nginx-proxy/concent-api-docs.tar

concent-api!dependencies:               \
    $(CONCENT_API_SOURCE)               \
    build/concent-api/requirements.lock \
    build/concent-api/testing.pref      \
    build/repositories/concent/

postgresql!dependencies: \
    $(POSTGRESQL_SOURCE)

IMAGE_DEPENDENCY_TARGETS := $(addsuffix !dependencies, $(IMAGE_TARGETS))

all: images

$(IMAGE_TARGETS): %: %!dependencies build/
    docker build --tag $(IMAGE_PREFIX)$@:$(IMAGE_VERSION) build/$@/
    docker tag $(IMAGE_PREFIX)$@:$(IMAGE_VERSION) $(IMAGE_PREFIX)$@:latest

.PHONY:                         \
    all                         \
    $(IMAGE_TARGETS)            \
    $(IMAGE_DEPENDENCY_TARGETS)
```

**Comment by michalpiasecki1** [comment]
superseded by https://github.com/golemfactory/concent-deployment/pull/211
